### PR TITLE
add code for app managed environment

### PR DESCRIPTION
**Pull request by kewalaka**
This code adds an initial implementation for the container app managed environment resource.

Current E2E tests cover default & an internal workload profile (I hope, if I merged the code into this repo right! 😂)

### Regarding the use of AzAPI

I think AzAPI is a good choice for this resource.
- aspects of Container Apps are moving faster than the AzureRM resource can keep up
- feedback on the latest [Terraform on Azure Community Call](https://youtu.be/xgj_yttYvBU?t=2371) re-enforced the support commitments around AzAPI and the validity for these use cases.  
- AVM modules by their nature are designed to abstract and hide essential complexity, a calling module consumer doesn't need to be concerned that AzAPI is used.
- There's no specific guidance from the AVM wiki I could find that suggests whether AzAPI is inappropriate (in fact, if i was in the product support group - i'd probably prefer AzAPI to be used because it looks more like Bicep 😂)

That said I note some disadvantages
- the means to manage changes to API versions is not clear
- the outputs may need some tidying to make them easier to consume
- workload profiles has just (in the last day or so), landed in AzureRM (still not released, last i checked).

### CODE REVIEW

**Review by segraef — COMMENTED** [review]
Regarding api versions, it's not mentioned somewhere you're right but also when using AzApi, this falls back to [SFR1](https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-sfr1---category-composition---preview-services) - _the latest “non-preview” version of the API SHOULD be used_.

**Comment by kewalaka** [comment]
> Regarding your statement "_the means to manage changes to API versions is not clear_", it's not mentioned somewhere you're right but also when using AzApi, this falls back to [SFR1](https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-sfr1---category-composition---preview-services) - _the latest “non-preview” version of the API SHOULD be used_.
> 
> I'm in conversation with the team about the current spec about API versions to be used and we'll update the documentation as required.

@segraef i was more meaning these sorts of concerns;

- how we would transition between API versions - would it simply be a breaking change requiring an updated API version, etc.  
-  whether we'd have to support multiple (which would be very messy), 
   - e.g. what if we _did_ want to offer access to preview features (noting the caveats about how their descriptions need to be decorated), but a customer didn't want to be using a preview API? 

Would be good to have some guidance on the appropriateness (or not) or azapi as well in general, I think.

File: main.tf [hunk 1]
```diff
@@ -1 +1,65 @@
-# TODO: insert resources here.
+data "azurerm_resource_group" "rg" {
+  name = var.resource_group_name
+}
```
[the remark this answers is not in the document]
Comment (kewalaka) [reply main.tf:3]: not sure where you mean sorry

File: variables.containerapps.env.tf [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+  type        = bool
+  description = "Restrict access to internal resources within VNet."
+  default     = false
+}
+
+variable "workload_profiles_enabled" {
+  type        = bool
+  description = "Whether to use workload profiles, this will create the default Consumption Plan, for dedicated plans use `workload_profiles`"
+  default     = false
+}
```
[the remark this answers is not in the document]
Comment (kewalaka) [reply variables.containerapps.env.tf]: not sure on this one either

**Comment by kewalaka** [comment]
ran the E2E tests in my fork:
https://github.com/kewalaka/terraform-azurerm-avm-res-app-managedenvironment/actions/runs/7124518461

File: terraform.tf [hunk 3]
```diff
@@ -10,5 +10,9 @@ terraform {
       source  = "hashicorp/random"
       version = ">= 3.5.0"
     }
+    azapi = {
+      source  = "Azure/azapi"
+      version = "1.9.0"
```
[the remark this answers is not in the document]
Comment (segraef) [reply terraform.tf]: Not sure why this was set as resolved.

**Review by segraef — COMMENTED** [review]
lgtm, let me check your workflow

File: variables.containerapps.env.tf
[the remark this answers is not in the document]
Comment (kewalaka) [reply variables.containerapps.env.tf:1]: @segraef  this has been removed, that is from the original commit, it was folded into variables prior to re-requesting this review

**Comment by kewalaka** [comment]
@segraef not sure if i'm doing something wrong, but your comments are from an old review.  If I look at the "files" section on this PR, I see what I expect.

it looks like this repo is missing the 'test' environment as well, needed for the e2e tests.  I don't have access to set that up.

**Comment by segraef** [comment]
> @segraef not sure if i'm doing something wrong, but your comments are from an old review. If I look at the "files" section on this PR, I see what I expect.
> 
> it looks like this repo is missing the 'test' environment as well, needed for the e2e tests. I don't have access to set that up.

You're not doing anything wrong, it was me looking at another commit or not having the page refreshed.

I'm currently setting up the e2e workflows and will run them then but for me/us your successful run is also fine for now.

**Comment by kewalaka** [comment]
Cool, in which case can this be merged?

**Comment by segraef** [comment]
Moving #1 to #3 to have e2e enabled.

**Pull request merged**